addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlCommunication.java

```diff
@@ -0,0 +1,479 @@
+                    nhcEventsRunning = false;
+                    // this is not an communication stop triggered from outside this runnable
+                    logger.warn("Niko Home Control: IO error in listener on thread {}", Thread.currentThread().getId());
+                    // the IO has stopped working, so we need to close cleanly and try to restart
+                    restartCommunication();
+                    return;
+                }
+            }
+
+            nhcEventsRunning = false;
+            // this is a stop from outside the runnable, so just log it and stop
+            logger.debug("Niko Home Control: event listener thread stopped on thread {}",
+                    Thread.currentThread().getId());
+        }
+
+    };
+
+    /**
+     * Method that interprets all feedback from Niko Home Control and calls appropriate handling methods.
+     *
+     * @param nhcMessage message read from Niko Home Control.
+     */
+    private void readMessage(String nhcMessage) {
+
+        logger.debug("Niko Home Control: received json {} on thread {}", nhcMessage, Thread.currentThread().getId());
+
+        try {
+            NhcMessageBase nhcMessageGson = this.gsonIn.fromJson(nhcMessage, NhcMessageBase.class);
+
+            String cmd = nhcMessageGson.getCmd();
+            String event = nhcMessageGson.getEvent();
+
+            if ("systeminfo".equals(cmd)) {
+                cmdSystemInfo(((NhcMessageMap) nhcMessageGson).getData());
+            } else if ("startevents".equals(cmd)) {
+                cmdStartEvents(((NhcMessageMap) nhcMessageGson).getData());
+            } else if ("listlocations".equals(cmd)) {
+                cmdListLocations(((NhcMessageListMap) nhcMessageGson).getData());
+            } else if ("listactions".equals(cmd)) {
+                cmdListActions(((NhcMessageListMap) nhcMessageGson).getData());
+            } else if ("executeactions".equals(cmd)) {
+                cmdExecuteActions(((NhcMessageMap) nhcMessageGson).getData());
+            } else if ("listactions".equals(event)) {
+                eventListActions(((NhcMessageListMap) nhcMessageGson).getData());
+            } else {
+                logger.debug("Niko Home Control: not acted on json {}", nhcMessage);
+            }
+        } catch (JsonParseException e) {
+            logger.debug("Niko Home Control: not acted on unsupported json {}", nhcMessage);
+        }
+
+    }
+
+    /**
+     * After setting up the communication with the Niko Home Control IP-interface, send all initialization messages.
+     * <p>
+     * Only at first initialization, also set the return values. Otherwise use the runnable to get updated values.
+     * While communication is set up for thermostats, tariff data and alarms, only info from locations and actions
+     * is used beyond this point in openHAB. All other elements are for future extensions.
+     *
+     * @throws IOException
+     */
+    private void initialize() throws IOException {
+
```

Comment:
I noticed that you very often start methods with an empty line - better remove them.